icu4c/source/i18n/formatted_string_builder.cpp

```diff
@@ -295,23 +295,19 @@ int32_t FormattedStringBuilder::prepareForInsertHelper(int32_t index, int32_t co
     }
     int32_t newZero;
     if (newLength > oldCapacity) {
-        int32_t newCapacity;
-        // Note: Value larger than the following may somehow cause
-        // malloc to get into strange state- it will return non-nullptr result
-        // and the early part of the memory is accessible but accessing the end
-        // of the returned memory will produce segment fault and crash inside
-        // the first uprv_memcpy2() below when running intltest
-        // format/ListFormatterTest/TestInt32Overflow -e
         if (newLength > INT32_MAX / 2) {
+            // We do not support more than 1G char16_t in this code because
+            // dealing with >2G *bytes* can cause subtle bugs.
             status = U_INPUT_TOO_LONG_ERROR;
             return -1;
         }
-        newCapacity = newLength * 2;
-        newZero = newLength / 2;
+        // Keep newCapacity also to at most 1G char16_t.
```

Comment:
FYI: The code does not actually do so right now... fodder for a follow-up ticket.